ccmbioinfo/lfs
# CARDS-1558: The generate_compose_yaml.py script should take into account the system's total memory and assign the appropriate memory limits for the Mongo and Java containers

This PR introduces memory limits on the _Wired Tiger_ cache used by the MongoDB data storage containers (shards / replicas) as well as the _memory heap_ of the CARDS Java process. These limits are enabled whenever the `generate_compose_yaml.py` script is instructed to provision a local MongoDB cluster as following the default memory limits can result in the server running out of memory as explained on https://phenotips.atlassian.net/browse/CARDS-1558.

Testing
---------

1. Build this `CARDS-1558` branch including a CARDS Docker image with `mvn clean install -Pdocker`.
2. `cd compose-cluster`
3. Clean up any prior Docker Compose configurations with `./cleanup.sh`.
4. Configure a simple CARDS local deployment that uses the file system for data persistence (`python3 generate_compose_yaml.py --oak_filesystem --dev_docker_image --composum`).
5. `docker-compose build`
6. `docker-compose up -d`
7. CARDS should soon be available at http://localhost:8080.
8. `docker-compose exec cardsinitial /bin/sh`
9. `ps -e`
10. The CARDS Java process should be listed as `java -Djdk.xml.entityExpansionLimit=0 -jar org.apache.sling.feature.launcher.jar -p .cards-data -u file:///root/.m2/.....`. Ensure that `-Xmx` is _not_ present.
11. Exit the `cardsinitial` container shell with `exit`.
12. Bring everything down with `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`.
13. Configure a CARDS local deployment that uses a MongoDB cluster for data persistence (`python3 generate_compose_yaml.py --shards 2 --replicas 3 --dev_docker_image --composum`).
14. `docker-compose build`
15. `docker-compose up -d`
16. Keep running `docker-compose ps` until you see the `compose-cluster_initializer_1` container show a _State_ of `Exit 0`,
17. CARDS should soon be available at http://localhost:8080.
18. `docker-compose exec cardsinitial /bin/sh`
19. `ps -e`
20. The CARDS Java process should be listed as `java -Djdk.xml.entityExpansionLimit=0 -Xmx{some number}m -jar org.apache.sling.feature.launcher.jar -p .cards-data -u file:///root/.m2/.....`. (eg. `java -Djdk.xml.entityExpansionLimit=0 -Xmx750m -jar org.apache.sling.feature.launcher.jar -p .cards-data -u file:///root/.m2/.....`) Ensure that `-Xmx` _is_ present.
21. Exit the `cardsinitial` container shell with `exit`.
22. `docker-compose exec s0r0 /bin/sh`
23. `ps aux`
24. Ensure that the `mongod` process is running with the `--wiredTigerCacheSizeGB` argument.
25. Exit the `s0r0` container shell with `exit`.
26. Bring everything down with `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`.
